einsum.lisp

```diff
@@ -0,0 +1,194 @@
+;;;; einsum.lisp
+;;;;
+;;;; Author: Erik Davis
+
+(defmacro einsum ((output-array . output-indices) &rest factors)
+  "Evaluates a sum of products of multidimensional arrays, summing
+over paired indices, storing the results in OUTPUT-ARRAY. 
+
+As an example, 
+
+ (einsum (A i j) (B i k) (C k j)) 
+
+results in the the updates
+ 
+ A[i,j] = \\sum_k B[i,k]C[k,j],
+
+which is equivalent to matrix multiplication.
+
+If OUTPUT-ARRAY is  _, then a newly allocated array will store the results, 
+e.g. 
+
+ (einsum (_ i) (A i j) (B j))
+
+computes the matrix-vector product AB and stores the results in a fresh array."
+  (apply #'%einsum
+           (unless (eq '_ output-array) output-array)
+           output-indices
+           factors))
+
+
+;;; A FACTOR is a list (A i1 ... ik) where - A is a symbol naming an
+;;; k-dimensional array - i1 ... ik are INDEX VARIABLES so that the
+;;; value of the factor, given a specific array and assignment of
+;;; index variables is A[i1,...,ik].
+
+(defun factor-aref-expr (factor)
```

Comment:
does `factor` include what is being dereferenced?

Reply:
I don't understand the question.

Reply:
The answer is "yes", based on the comment block above. It might be nice to move that into a docstring.